backend/cmm_helpers.ml

```diff
@@ -1615,71 +1668,28 @@ let make_alloc_generic ?(scannable_prefix = Scan_all) ~mode set_fn dbg tag
               },
             List.map (fun arg -> Cconst_int (arg, dbg)) caml_alloc_args,
             dbg ),
-        fill_fields 0 args )
-
-let addr_array_init arr ofs newval dbg =
-  Cop
-    ( Cextcall
-        { func = "caml_initialize";
-          ty = typ_void;
-          alloc = false;
-          builtin = false;
-          returns = true;
-          effects = Arbitrary_effects;
-          coeffects = Has_coeffects;
-          ty_args = []
-        },
-      [array_indexing log2_size_addr arr ofs dbg; newval],
-      dbg )
+        fill_fields 0 args args_memory_chunks )
 
 let make_alloc ~mode dbg ~tag args =
-  make_alloc_generic ~mode
-    (fun _ arr ofs newval dbg -> addr_array_init arr ofs newval dbg)
-    dbg tag (List.length args) args
+  make_alloc_generic ~scannable_prefix:Scan_all ~mode dbg tag (List.length args)
+    args
+    (List.map (fun _ -> Word_val) args)
 
 let make_float_alloc ~mode dbg ~tag args =
-  make_alloc_generic ~mode
-    (fun _ -> float_array_set)
-    dbg tag
+  make_alloc_generic ~scannable_prefix:Scan_all ~mode dbg tag
     (List.length args * size_float / size_addr)
     args
+    (List.map (fun _ -> Double) args)
 
-module Flat_suffix_element = struct
-  type t =
-    | Tagged_immediate
-    | Naked_float
-    | Naked_float32
-    | Naked_int32
-    | Naked_int64_or_nativeint
-end
+let make_closure_alloc ~mode dbg ~tag args args_memory_chunks =
+  let size = mixed_block_size args_memory_chunks in
+  make_alloc_generic ~scannable_prefix:Scan_all ~mode dbg tag size args
```

Comment:
It's a bit strange to pass `Scan_all` since the arguments can still have an unboxed prefix. Can we rename it?

Reply:
We certainly can change the name of these constructors, maybe something along the lines of `Regular_block` and `Mixed block` ? (contrary to what one might think, a closure, even one containing unboxed values, is a regular block from that point of view, since they use the startenv trick to avoid scanning the unboxed values).

Reply:
Sounds good

Reply:
We should probably change the name of the label too, in that case.

Reply:
I've done the renaming.